Test tip model

MrBeardface opened the pull request.

MrBeardface committed:
Test the Tip model

- Test #title_and_username should be valid if slug is equal to
  parameratized version
- Test #should_generate_friendly_id? is valid if tip was created more
  then one day ago.
- Test #should_generate_friendly_id? is invalid if tip was created less
  then one day ago or if tip hasn't been created yet.
- Test #title_must_be_unique_for_user is valid if tip slug doesn't
  already exist.
- Test #title_must_be_unique_for_user is invalid if tip already exists.

MrBeardface committed:
Major issues with MySQL DB

MrBeardface committed:
Fixed db

MrBeardface committed:
Update Tip specs

- Redo test expectations after reading Rails 4 Test Prescriptions

MrBeardface committed:
Redo tests for the Tip model

- Still need to write correct spec for "adds correct error message", it
  passes, but is currently testing that [:title] array isn't empty, not
	that its adding the correct error message.
- Fixed spelling issue in Tip model

MrBeardface committed:
Fix spec for Tip model

- Fix "adds correct error message" spec for Tip. It now tests that the
  error message is added to the [:title] array when Tip already exists.

MrBeardface committed:
Clean up tests

- Clean up #should_generate_new_friendly_id? to test if tip is
  persisted.
- Clean up relationship tests for user and bookmarks.

MrBeardface committed:
Remove unnecessary tests for associations

- Remove two tests that evaluate association between user and bookmarks.

MrBeardface committed:
Adjust Tip model specs and factories

- Update Factory girl data for Tip and User.
- Test whether tips are in order by time created.

MrBeardface committed:
Change tests to test all of Tip model

- Adjust factories for Tip
- Test pygmentize_and_cache_body method

MrBeardface committed:
Update describe and context blocks

- Update describe and context blocks to more accurately describe what
  tests are doing.

MrBeardface committed:
Merge branch 'master' of github.com:asm-products/tips-on-rails into Test-Tip-Model

Conflicts:
	spec/factories/tip.rb

ohlincik commented on spec/models/tip_spec.rb:
```diff
@@ -2,51 +2,158 @@
+    end
   end
 
   describe "with blank title" do
```
Move all the validation tests up so they are 'grouped' with the other validations. Also, use the shoulda matchers that allow you to test the validation conditions without having to do the individual setups for each test. Check how I implemented these in the User model.

MrBeardface replied:
I will look at your code and change the validations.

ohlincik commented on spec/models/tip_spec.rb:
```diff
@@ -2,51 +2,158 @@
-    it { should_not be_valid }
+  describe "#title_and_username" do 
+    context "method creates string that" do
+      it "should be equal" do
```
Why do you need to test this in a loop? Wouldn't it be sufficient to test that this works for a single tip?

MrBeardface replied:
Failure/Error: expect(tip.title_and_username).to eq("#{tip.title} by #{user.username}")

       expected: "Tip title 16 by user_17"
            got: "Tip title 16 by user_16"

       (compared using ==)

Due to the way we've set up the test database, it can't find the correct user when the test is run. To ensure that the right user is found, it's probably best to loop through each one.

ohlincik replied:
You don't need to do anything with database here at all. You can use FactoryGirl to build_stubbed user and then build_stubbed tip for that user. Then you can test that the method generates the right string.

ohlincik commented on spec/models/tip_spec.rb:
```diff
@@ -2,51 +2,158 @@
+    end
+  end
+
+  describe "#pygmentize_and_cache_body" do
```
I don't think that this test actually tests the right thing.

MrBeardface replied:
What do you think that it is testing?

MrBeardface replied:
Please a bit more specific about the method you don't think is being tested right. It's hard to see on my end which method you're referring to.

ohlincik commented on spec/models/tip_spec.rb:
```diff
@@ -2,51 +2,158 @@
+    end
+  end
+
+  describe "#should_generate_friendly_id?" do
```
I don't think these tests are right. You should have the need to use any conditionals here. That's what the actual method should do. You should just set the created_at to known values and test for the method to return true or false.

MrBeardface replied:
For - describe "#should_generate_friendly_id?" - the should_generate_friendly_id? is based upon whether the tip was created_at? more then a day ago. The method doesn't just know to generate a new id, it has to be told based upon what the condition tells it to do.

ohlincik commented on spec/models/tip_spec.rb:
```diff
@@ -2,51 +2,158 @@
+    end
+  end
+           
+  describe "#title_must_be_unique_for_user" do
```
Again, you don't need a loop here to test whether the method works properly.

ohlincik commented on spec/models/tip_spec.rb:
```diff
@@ -2,51 +2,158 @@
+
+  describe "#pygmentize_and_cache_body" do
+    it "should render the tip body" do
+      expect(MarkDownRenderer.new.pygmentize_and_render(tip.body)).to eq("<p>#{tip.body}</p>\n")
```
I believe that in order to test a method you actually need to use that method...

MrBeardface committed:
Updates Tests on Tip Model

- Change descriptions of multiple tests to accruately describe them.
- Update how validations are tested.
- Update #pygmentize_and_cache_body to test the method, not contents of
  method.
- Update #should_generate_new_friendly_id? to test if tip is persisted.
